### PR TITLE
Testnet deployment flow

### PR DESCRIPTION
**Pull request by loongy**
**Description**

This PR brings the `master` branch back up-to-date with the `develop` branch, since the two have diverged quite a bit in the last month. This PR is being made in preparation for F∅, allowing us to use our new development/deployment work flow for our internal testnet environments vs. public testnets.

**Motivation**

Internally, we would like to use various branches for deployments to various internal, and public, testnets. This PR is a step towards to testing and utilizing that control flow.

**Unresolved Issues**

https://github.com/republicprotocol/republic-go/pull/46 has been merged through triage and still needs to be addressed in a new PR. This PR does not attempt to be a replacement for the work proposed by https://github.com/republicprotocol/republic-go/pull/46.

### CODE REVIEW

**Comment by tok-kkk** [comment]
Do we want to remove some of the test logs for master ? i.e. logs of different stages in computer

**Comment by loongy** [comment]
This CI is now passing with a couple minor issues fixed. Waiting for https://github.com/republicprotocol/republic-go/pull/44 so that our logging is back up to date, and it should also improve test coverage.

File: .travis.yml [hunk 1]
```diff
@@ -3,29 +3,24 @@ language: go
...
-  - golint -set_exit_status `go list ./... | grep -Ev "(stackint/asm|vendor)"`
-  - go run cmd/testnetwork/main.go -sleep 10 &
-  - sleep 10
-  - goverage -v -coverprofile=coverprofile.out ./...
-
-after_success:
-  - sed -i '/rpc.pb.go/d' coverprofile.out
+  # - go vet ./...
+  # - golint -set_exit_status `go list ./... | grep -Ev "(stackint/asm|vendor)"`
+  # - golint `go list ./... | grep -Ev "(stackint/asm|vendor)"`
```
Comment (noahingham) [inline .travis.yml]: `go vet` and `golint` can be re-enabled now

Comment (loongy) [reply .travis.yml]: Good idea, let's re-enable this once everything else is ready to go. Then we can go through and document types and functions in bulk.

File: logger/logger_test.go [hunk 2]
```diff
@@ -1,9 +1,875 @@
...
+	log := Log{}
+	if err != nil {
+		return log, err
+	}
+	// Marshal the file back to a log
+	json.NewDecoder(file).Decode(&log)
+	return log, nil
+}
+
+func resetTmp() error {
```
Comment (noahingham) [inline logger/logger_test.go]: `resetTmp` doesn't appear to be used anywhere

File: ome/confirmer_test.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+func randomCoExp() order.CoExp {
+	co := uint64(rand.Intn(1999) + 1)
+	exp := uint64(rand.Intn(25))
+	return order.CoExp{
+		Co:  co,
+		Exp: exp,
+	}
+}
+
+func ganacheConfig() ethereum.Config {
```
Comment (noahingham) [inline ome/confirmer_test.go]: `ganacheConfig` doesn't appear to be used anywhere

File: http/http.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/republicprotocol/republic-go/http/adapter"
+	"github.com/rs/cors"
+)
+
+const reset = "\x1b[0m"
```
Comment (noahingham) [inline http/http.go]: `reset` doesn't appear to be used anywhere

File: blockchain/ethereum/ethereum.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+			config.RenExAccountsAddress = RenExAccountsAddressOnKovan.String()
+		case NetworkRopsten:
+			panic("not deployed yet")
+		default:
+			return Conn{}, fmt.Errorf("cannot connect to %s: unsupported", config.Network)
+		}
+	}
+	// Fixme: hardcode the renex contract for now,
+	// need to update the address in therir config files
+	config.RenExAccountsAddress = RenExAccountsAddressOnKovan.String()
```
Comment (tok-kkk) [inline blockchain/ethereum/ethereum.go]: Do we want to have all the contract addresses in the config file or coed in the darknode ?

File: grpc/grpc_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+func TestRpc(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "gRPC Suite")
+}
+
+// mockVerifier is a simple implementation of the crypto.Verifier interface
+// that accepts all signatures.
+type mockVerifier struct {
```
Comment (noahingham) [inline grpc/grpc_suite_test.go]: `mockVerifier` and `mockSigner` don't appear to be used anywhere

File: blockchain/ethereum/ledger/ledger.go [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+		transactOpts: transactOpts,
+		callOpts:     callOpts,
+		binding:      contract,
+		address:      common.HexToAddress(conn.Config.RenLedgerAddress),
+	}, nil
+}
+
+func (ledger *RenLedgerContract) OpenBuyOrder(signature [65]byte, id order.ID) error {
+
+	ledger.transactOpts.GasPrice = big.NewInt(int64(5000000000))
```
Comment (tok-kkk) [inline blockchain/ethereum/ledger/ledger.go]: Should we allow them to set the gas price? Not sure if 5 gwei is good for all case.

File: blockchain/ethereum/ethereum.go (hunk 5 above)
Comment (loongy) [reply blockchain/ethereum/ethereum.go]: It was agreed that having them directly programmed into the Darknodes was better for rolling out updates. Going through and changing that now.

File: blockchain/ethereum/ledger/ledger.go (hunk 7 above)
Comment (loongy) [reply blockchain/ethereum/ledger/ledger.go]: I think having a default is a good idea, but 5GWEI is probably too low. Review task has been added.

File: dispatch/error_test.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+			errCh <- err1
+			errCh <- err2
+			errCh <- err3
+
+			predicate := func(err error) bool {
+				if len(err.Error()) == 2 {
+					return true
+				}
+				return false
```
Comment (noahingham) [inline dispatch/error_test.go]: This can be simplified to `return len(err.Error()) == 2`

File: blockchain/test/test.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
+package test
```
Comment (tok-kkk) [inline blockchain/test/test.go]: This file should be moved to the testutils package.

File: cal/darkpool_accounts.go [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package cal
+
+import "github.com/republicprotocol/republic-go/order"
+
+type DarkpoolAccounts interface {
+
+	// Settle the order pair which gets confirmed by the RenLedger
+	Settle(buy order.Order, sell order.Order) error
+
+	// Balance returns how many token the trader
```
Comment (tok-kkk) [inline cal/darkpool_accounts.go]: `Balance returns how many token the trader has`

File: blockchain/test/ganache/ganache.go [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+	globalGanacheCmd.Process.Signal(syscall.SIGTERM)
+	globalGanacheCmd.Wait()
+}
+
+// StopOnInterrupt will stop Ganache upon receiving receiving a interrupt signal
+func StopOnInterrupt() {
+	signals := make(chan os.Signal, 1)
+	defer close(signals)
+
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGKILL, syscall.SIGTERM)
```
Comment (noahingham) [inline blockchain/test/ganache/ganache.go]: syscall.SIGKILL should be removed as it can't be caught

File: cmd/analysis/leveldb/main.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+		log.Fatalf("cannot open leveldb: %v", err)
+	}
+
+	iter := orderFragments.NewIterator(nil, nil)
+	fragmentsCount := 0
+	for iter.Next() {
+		// Remember that the contents of the returned slice should not be modified, and
+		// only valid until the next call to Next.
+		key := iter.Key()
+		orderID := common.ToHex(key)
```
Comment (noahingham) [inline cmd/analysis/leveldb/main.go]: `ToHex` has the warning `Deprecated: use hexutil.Encode instead` (applies to two other uses of `ToHex`)

File: cmd/workspace/workspace.go [hunk 13]
```diff
@@ -0,0 +1,61 @@
+package main
+
```
Comment (tok-kkk) [inline cmd/workspace/workspace.go]: I think this file can be removed.

File: cmd/registry/registry.go [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+
+		if isRegistered {
+			_, err = registry.Deregister(address.ID())
+			if err != nil {
+				return fmt.Errorf("[%v] %sCouldn't deregister node%s: %v\n", address, red, reset, err)
+			} else {
+				log.Printf("[%v] %sNode will be deregistered next epoch%s\n", address, green, reset)
+			}
+		} else if isRegistered {
+			log.Printf("[%v] %sNode already registered%s\n", address, yellow, reset)
```
Comment (noahingham) [inline cmd/registry/registry.go]: `else if isRegistered` should either be `!isRegistered` or just an `else`, and the log should say already **deregistered**

File: cmd/test.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+		// Check if node has already been registered
+		isRegistered, err := registry.IsRegistered(addr)
+		if err != nil {
+			log.Fatalf("[%v] %sCouldn't check node's registration%s: %v\n", address, red, reset, err)
+		}
+
+		// Register the node if not registered
+		if isRegistered {
+			_, err = registry.Deregister(addr.ID())
```
Comment (noahingham) [inline cmd/test.go]: This should check `!isRegistered` and then call `Register`.

File: order/value.go [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+
+// NewCoExp returns a CoExp with the given coefficient and exponent.
+func NewCoExp(co uint64, exp uint64) CoExp {
+	return CoExp{
+		Co:  co,
+		Exp: exp,
+	}
+}
+
+func (val CoExp) ToFloat() float64 {
```
Comment (tok-kkk) [inline order/value.go]: This seems to be wrong.

File: grpc/stream.go [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+	str.sendMu.Lock()
+	str.recvMu.Lock()
+	defer str.sendMu.Unlock()
+	defer str.recvMu.Unlock()
+
+	if str.done == nil {
+		return nil
+	}
+	close(str.done)
+	str.done = nil
```
Comment (noahingham) [inline grpc/stream.go]: `str.done` is being assigned to for a non-pointer `str`

File: blockchain/test/test.go (hunk 9 above)
Comment (loongy) [reply blockchain/test/test.go]: Review task has been added.

**Pull request merged**